<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Microarray-analysis by bioinformatics-core-shared-training</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Microarray-analysis</h1>
      <h2 class="project-tagline">Materials on the analysis of microarray expression data; focus on re-analysis of public data</h2>
      <a href="https://github.com/bioinformatics-core-shared-training/microarray-analysis" class="btn">View on GitHub</a>
      <a href="https://github.com/bioinformatics-core-shared-training/microarray-analysis/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/bioinformatics-core-shared-training/microarray-analysis/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h3>
<a id="analysis-of-publicly-available-microarray-data" class="anchor" href="#analysis-of-publicly-available-microarray-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Analysis of publicly-available Microarray data</h3>

<p>Although microarrays have been superseded by high-throughput sequencing technologies for gene expression profiling, years of experience gained from analysing microarray data has led to a variety of analysis techniques and datasets that can be exploited in other contexts. In this course, we will focus on retrieving and exploring microarray data from public repositories such as Gene Expression Omnibus (GEO). </p>

<h2>
<a id="aims-during-the-course-you-will-learn-about" class="anchor" href="#aims-during-the-course-you-will-learn-about" aria-hidden="true"><span class="octicon octicon-link"></span></a>Aims: During the course you will learn about</h2>

<ul>
<li>Exploratory data analysis techniques for high-throughput data</li>
<li>Workflows for the analysis of Illumina and Affymetrix gene expression data</li>
<li>Normalisation of gene expression data</li>
<li>Differential expression (DE) analysis using linear-modelling techniques</li>
<li>Importing data from GEO into R</li>
<li>Principal Components Analysis and hierarchical clustering of gene expression data </li>
</ul>

<h2>
<a id="objectives-after-this-course-you-should-be-able-to" class="anchor" href="#objectives-after-this-course-you-should-be-able-to" aria-hidden="true"><span class="octicon octicon-link"></span></a>Objectives: After this course you should be able to</h2>

<ul>
<li>Import gene expression datasets from GEO into R</li>
<li>Assess the quality of a dataset in a repository</li>
<li>Identify, and correct for, batch effects</li>
<li>Perform a standard DE analysis to get a ranked list of genes</li>
<li>Use un-supervised methods to explore a dataset</li>
<li>Interrogate particular genes of interest</li>
</ul>

<h2>
<a id="prerequisites" class="anchor" href="#prerequisites" aria-hidden="true"><span class="octicon octicon-link"></span></a>Prerequisites</h2>

<ul>
<li>A very basic knowledge of UNIX would be an advantage, but nothing will be assumed and extremely little will be required</li>
<li>Attendees should be comfortable with using the R statistical language to read and manipulate data, and produce simple graphs</li>
</ul>

<h3>
<a id="software" class="anchor" href="#software" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to Run the course.</h3>
We recommend using <a href="www.rstudio.com">RStudio</a> for the practicals along with <a href="https://cran.r-project.org/">R version 3.2.3</a>
<p>
Download the materials from this repository and install the required R and Bioconductor packages from within RStudio. This may take several minutes.
<pre class="input"><code>source("http://www.bioconductor.org/biocLite.R")
biocLite(c("limma","affy","affyPLM","beadarray","GEOquery","genefilter","illuminaHumanv3.db","cluster",
           "ggplot2","GOstats","breastCancerVDX","breastCancerTRANSBIG","breastCancerNKI","pamr","survival",
           "estrogen","ArrayExpress","RColorBrewer","arrayQualityMetrics","hgu95av2cdf","hgu95av2.db","org.Hs.eg.db","wakefield"))
</code></pre>


<h2>
<a id="day-1" class="anchor" href="#day-1" aria-hidden="true"><span class="octicon octicon-link"></span></a>Day 1</h2>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/intro.html">Introduction to Microarray technologies</a> </li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/recap.html">Introductory Practical</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/affymetrix.html">A workflow for the analysis of Affymetrix arrays </a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/illumina.html">A workflow for the analysis of Illumina arrays (Optional)</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/DesignLinearModels.pdf">Linear Models</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/public-data.html">Getting data from public repositories</a> </li>
<h2>
<a id="day-2" class="anchor" href="#day-2" aria-hidden="true"><span class="octicon octicon-link"></span></a>Day 2</h2>

<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/de-tutorial.html">Differential Expression tutorial</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/Clustering_Classification_Survival.pdf">Downstream methods Lecture</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/clus-and-class.html">Clustering and Classification tutorial</a></li>
<li><a target="_blank" href="https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/enrichment.html">Enrichment and ontologies tutorial</a></li>
      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/bioinformatics-core-shared-training/microarray-analysis">Microarray-analysis</a> is maintained by <a href="https://github.com/bioinformatics-core-shared-training">bioinformatics-core-shared-training</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
