---
title: "Differential Expression Tutorial"
author: "Your Name Here"
date: '16th Feb. 2016'
output: html_document
---

# Importing the data


```{r}
library(GEOquery)
library(limma)
```


```{r}
library(GEOquery)
library(limma)
url <- "ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SeriesMatrix/GSE33126/GSE33126_series_matrix.txt.gz"
filenm <- "data/GSE33126_series_matrix.txt.gz"
if(!file.exists(filenm)) download.file(url, destfile=filenm)
gse <- getGEO(filename=filenm)
gse
```

```{r}
head(exprs(gse))
```


******
##Q  Do the data look to be normalised?

******

```{r}
exprs(gse) <- log2(exprs(gse))
boxplot(exprs(gse),outline=FALSE)
```

Inspect the clinical variables

******
##Q What column has the information about whether is sample is a tumour or normal?
##Q How many do we have in each group?

******
```{r}
```

# Simple t-tests


```{r}
library(genefilter)
pd <- pData(gse)
SampleGroup <- pd$source_name_ch1
destats <- rowttests(exprs(gse),SampleGroup)
head(destats)
```

- Can we remember the drawbacks of this approach?


# The linear model approach




```{r}
design <- model.matrix(~0+SampleGroup)
design
colnames(design) <- c("Normal","Tumour")
```

```{r}
design <- matrix(nrow=18,ncol=2)
design[,1] <- c(0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1)
design[,2] <- c(1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0)
colnames(design) <- c("Normal","Tumour")
```


```{r}
design[,1] <- rep(c(0,1),by=9)
design[,2] <- rep(c(1,0),by=9)
```


```{r}
library(genefilter)
gse.expFilt <- varFilter(gse)
gse.expFilt
```


```{r}
fit <- lmFit(exprs(gse.expFilt), design)
names(fit)
dim(fit$coefficients)
head(fit$coefficients)
```


******

##Q. What is the interpretation of the `coefficients` item in the output

******



```{r}
contrasts <- makeContrasts(Tumour - Normal, levels=design)
fit2 <- contrasts.fit(fit, contrasts)
head(fit2$coeff)
```


```{r}
fit2 <- eBayes(fit2)
fit2
```


```{r}
topTable(fit2)
```


# Including annotation

```{r}
anno <- fData(gse.expFilt)
head(anno)[,1:5]
anno <- anno[,c("Symbol","Entrez_Gene_ID","Chromosome","Cytoband")]
fit2$genes <- anno
topTable(fit2)
```

# Diagnostic checks


```{r}
dotchart(exprs(gse)["ILMN_1704294",])
```

```{r}
boxplot(exprs(gse)["ILMN_1704294",]~SampleGroup)
```


```{r}
decideTests(fit2)
table(decideTests(fit2))
```

```{r}
volcanoplot(fit2)
```


```{r}
volcanoplot(fit2,highlight=10,names = fit2$genes$"Symbol")
```


# Exporting the results

```{r}
write.fit(fit2, file = "de-results.txt",adjust="BH")
```

# Getting the results for a particular gene


```{r}
testResults <- topTable(fit2, number=nrow(fit2))
```


```{r}
testResults[which(testResults$Symbol == "OCIAD2"),]
```



```{r}
mylist <- c("LOC441066","ARF3","FMNL3","CSF1R","XLKD1","TTRAP","DMWD","SYNPO2L","PILRB","LAMP3")

testResults[which(testResults$Symbol %in% mylist),]

```

